src/keyboard/mod.rs

```diff
@@ -0,0 +1,258 @@
+    // TODO: better error handling
+    let mut keymap_file = tempfile().unwrap();
+    let keymap_data = internal.keymap.get_as_string(xkb::KEYMAP_FORMAT_TEXT_V1);
+    keymap_file.write_all(keymap_data.as_bytes()).unwrap();
+    keymap_file.flush().unwrap();
+
+    Ok(KbdHandle {
+           internal: Arc::new((Mutex::new(internal), (keymap_file, keymap_data.as_bytes().len() as u32))),
+       })
+}
+
+/// An handle to a keyboard handler
+///
+/// It can be cloned and all clones manipulate the same internal state. Clones
+/// can also be sent across threads.
+///
+/// See module-level documentation for details of its use.
+#[derive(Clone)]
+pub struct KbdHandle {
+    internal: Arc<(Mutex<KbdInternal>, (::std::fs::File, u32))>,
+}
+
+impl KbdHandle {
+    /// Handle a keystroke
+    ///
+    /// All keystrokes from the input backend should be fed _in order_ to this method of the
+    /// keyboard handler. It will internally track the state of the keymap.
+    ///
+    /// The `filter` argument is expected to be a closure which will peek at the generated input
+    /// as interpreted by the keymap befor it is forwarded to the focused client. If this closure
+    /// returns false, the input will not be sent to the client. This mechanism can be used to
+    /// implement compositor-level key bindings for example.
+    ///
+    /// The module `smithay::keyboard::keysyms` exposes definitions of all possible keysyms
+    /// to be compared against. This includes non-characted keysyms, such as XF86 special keys.
+    pub fn input<F>(&self, keycode: u32, state: KeyState, serial: u32, filter: F)
+        where F: FnOnce(&ModifiersState, Keysym) -> bool
+    {
+        let mut guard = self.internal.0.lock().unwrap();
+        let mods_changed = guard.key_input(keycode, state);
+
+        if !filter(&guard.mods_state, guard.state.key_get_one_sym(keycode)) {
+            // the filter returned false, we do not forward to client
+            return;
+        }
+
+        // forward to client if no keybinding is triggered
+        if let Some((_, ref kbd)) = guard.focus {
+            if mods_changed {
+                let (dep, la, lo, gr) = guard.serialize_modifiers();
+                kbd.modifiers(serial, dep, la, lo, gr);
+            }
+            let wl_state = match state {
+                KeyState::Pressed => wl_keyboard::KeyState::Pressed,
+                KeyState::Released => wl_keyboard::KeyState::Released,
+            };
+            kbd.key(serial, 0, keycode, wl_state);
+        }
+    }
+
+    /// Set the current focus of this keyboard
+    ///
+    /// Any previous focus will be sent a `wl_keyboard::leave` event, and if the new focus
+    /// is not `None`, a `wl_keyboard::enter` event will be sent.
+    pub fn set_focus(&self, focus: Option<(wl_surface::WlSurface, wl_keyboard::WlKeyboard)>, serial: u32) {
+        // TODO: check surface and keyboard are from the same client
```

Comment:
The concept of "keyboard focus" is really part of the core wayland protocol, and a very classic behavior. Most wayland clients will expect this setup (for example, an app with multiple windows will likely want to know in which of them the user is typing).

If a compositor does not want to do it that way, well, they'd just no use this helper. That the whole philosophy I wand to put in smithay: provide helpers that are good for 90% of the use-cases, and the compositors built on top of it use only what they want/need. :)